src/main/java/com/google/research/bleth/services/DatabaseService.java

```diff
@@ -0,0 +1,223 @@
+package com.google.research.bleth.services;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.repackaged.com.google.gson.Gson;
+import com.google.common.collect.ArrayTable;
+import com.google.common.collect.Multimap;
+import com.google.research.bleth.exceptions.BoardStateAlreadyExistsException;
+import com.google.research.bleth.exceptions.ExceedingRoundException;
+import com.google.research.bleth.simulator.AbstractSimulation;
+import com.google.research.bleth.simulator.Board;
+import com.google.research.bleth.simulator.IAgent;
+import com.google.research.bleth.simulator.Location;
+
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * DatabaseService is a singleton class wrapping the application's datastore instance.
+ * DatabaseService provides a global point of access to any read/write operation using the datastore instance.
+ *
+ * DatabaseService provides API for:
+ * - Writing simulation metadata by the builder responsible for the simulation construction (which returns the unique id
+ * assigned to the simulation, encoded to a string).
+ * - Writing a board state to the db.
+ * - Reading a board state from the db (which returns null if provided round exceeds simulation's max number of round,
+ * or a json string encoding the board state).
+ */
+public class DatabaseService {
+
+    private static DatabaseService instance = null;
+    private DatastoreService datastore;
+    private Gson gson = new Gson();
+
+    /**
+     * Static method to create an instance of the DatabaseService class.
+     * @return a DatabaseService class instance.
+     */
+    public static DatabaseService getInstance() {
+        if (instance == null) {
+            instance = new DatabaseService();
+        }
+        return instance;
+    }
+
+    /**
+     * Write an entity providing simulation metadata.
+     * @param builder is the builder responsible for the simulation construction.
+     * @return the new simulation id (encoded to string).
+     */
+    public String writeMetadata(AbstractSimulation.Builder builder) {
+        // Create new entity.
+        Entity entity = new Entity("Simulation");
+
+        // Set properties.
+        entity.setProperty("type", builder.getSimulationType());
+        entity.setProperty("roundsNum", builder.getMaxNumberOfRounds());
+        entity.setProperty("beaconsNum", builder.getBeaconsNum());
+        entity.setProperty("observersNum", builder.getObserversNum());
+        entity.setProperty("rowNum", builder.getRowNum());
+        entity.setProperty("colNum", builder.getColNum());
```

Comment:
the database service seems to be aware of internal structure of the simulation object (e.g., why should it care about the movement strategy?). this creates a dependency in the wrong direction.

Reply:
The idea here was that the `builder` will pass himself to the `writeSimulationMetadata()` method inside its `build()` method. Something like:

```java
public TracingSimulation build() {
    this.id = DatabaseService.getInstance().writeSimulationMetadata(this);
    return new TracingSimulation(this);
}
```

Alternatively we can modify `writeSimulationMetadata()` to recieve multiple arguments instead of a builder, and then our `build()` method will look like:

```java
public TracingSimulation build() {
    this.id = DatabaseService.getInstance().writeSimulationMetadata(this.rowNum, this.colNum, /** lots of other attributes */);
    return new TracingSimulation(this);
}
```

But this smells bad as well. I guess what we're talking about is having some other object in the middle, which specifies both the properties names and the values to be written to the db. Let's say we're using some kind of a HashMap:

```java
public TracingSimulation build() {
    HashMap<String, Double> metadataEntity = new HashMap<>();
    metadataEntity.put("rowNum", this.rowNum);
    metadataEntity.put("colNum", this.colNum);
    /** lots of other attributes */
    this.id = DatabaseService.getInstance().write(metadataEntity);
    return new TracingSimulation(this);
}
```

And this might be better, because now the `write()` method can be similar for all entities we're writing. But someone must create this HashMap (or whatever we will work with) everytime, including specifing the properties names. 

Did you think of something else? @schainmariano

Reply:
Also, the reason we're writing the Simulation's strategies as part of the metadata is to able to provide answers for questions like "what was the average result in all simulations where the beacons moved **randomally** and the observers where **static**?"